### PR TITLE
PDP publish workflow

### LINKED ISSUE #1178: Product detail page editing workflow improvements
Simple publishing workflow for products and variants.

Goals:
- [ ] Allow the admin users to edit a product without having changes showing up immediately for all users.
- [ ] Save any changes to a product as a draft revision in the `Revisions` Collection.
- [ ] Use a "Publish Changes" button to publish all draft changes to a product and all variants with a draft.

### PR DESCRIPTION
**Pull request by mikemurray**
Resolves #1178 

Basic draft / revision workflow for the Product detail page and products in the product grid.

- [ ] Description explains the issue / use-case resolved
- [ ] Only contains code directly related to the issue
- [ ] Has tests.
- [ ] Has docs.
- [ ] Passes all tests
- [ ] Has been linted and follows the style guide

### CODE REVIEW

File: imports/plugins/core/revisions/server/publications.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { Products, Revisions } from "/lib/collections";
```
Comment (aaronjudd) [inline imports/plugins/core/revisions/server/publications.js]: Is there a specific reason that Products is imported here?

File: lib/collections/schemas/revisions.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+  workflow: {
+    type: Workflow,
+    optional: false
+  },
+
+  documentId: {
+    type: String,
+    label: "Reference Document Id"
+  },
```
Comment (aaronjudd) [inline lib/collections/schemas/revisions.js]: Perhaps "index: true" would be appropriate here?

File: imports/plugins/core/revisions/server/publications.js (hunk 1 above)
Comment (mikemurray) [inline imports/plugins/core/revisions/server/publications.js]: Probably none as I went with doing the revision publication from the products publication.

This file might need a rewrite or removal.

**Comment by aaronjudd** [comment]
Should I be seeing empty change lines when adding details? This occurs after the first time you publish changes with details, prior to publishing you see [Object,Object] (in the change column).
basic_reaction_product

**Comment by aaronjudd** [comment]
Some ui thoughts.. for now or in the future, having warnings display with the publish changes could be helpful.

As a user I felt it could be a little unclear what is published or not, blockers/warnings would be good, but also I'm not sure the color coding on the table tells me clearly enough what the current status of the live product is.

Btw - the pencil icon confused me there.. I was hoping I could edit that change. 

ctlndpqiram9ilmkm

**Comment by mikemurray** [comment]
I think each field / section needs some indication that its been edited, not just that diff in the sidebar, which is a pretty rough summary at the moment. Perhaps thats a colored badge or something else.

**Comment by aaronjudd** [comment]
I was scrolled down in a smaller window and watching the publish view while testing, so I was oblivious to the warnings at the top, so I guess it's always been an issue but now we have somewhere to put them.

File: imports/plugins/included/product-variant/client/templates/products/productSettings/productSettings.js [hunk 3]
```diff
@@ -17,19 +19,34 @@ Template.productSettings.onCreated(function () {
...
       });
-
+      this.state.set("productIds", productIds);
       const products = Products.find({
         _id: {
           $in: productIds
         }
-      }).fetch();
+      }).map((product) => {
+        console.log(product);
```
Comment (aaronjudd) [inline imports/plugins/included/product-variant/client/templates/products/productSettings/productSettings.js]: Errant console log is spitting the product object out on the browser.

**Comment by aaronjudd** [comment]
Seems to be incompatible with current product structure, if I switch to this branch with existing product data, I get "No Products Found".  This would be a blocker, I think.

File: imports/plugins/core/revisions/server/startup.js [hunk 4]
```diff
@@ -0,0 +1,175 @@
+import { Products, Revisions } from "/lib/collections";
+import { Logger } from "/server/api";
+import { diff } from "deep-diff";
+
+
+Products.before.insert((userId, product) => {
```
Comment (aaronjudd) [inline imports/plugins/core/revisions/server/startup.js]: I think hooks should move to `server/hooks/hooks.js` for more consistent implementation of hooks.  Is there a specific reason it's in startup.js?

**Comment by zenweasel** [comment]
This has been moved to v16 so let's reopen when we are ready to review then.

**Pull request closed without merging**